### ISSUE TITLE
Not able to get the Env variable in the Jupyter Notebook

### ISSUE THREAD

**Issue by IamViditAgarwal** [jupyter/notebook#4600]
Hi.
I m running a Jupyter Notebook in a docker container.
I am trying to set some ENVIRONMENT variables based on the user after the Jupyter Notebook is opened.
Although I am able to set the variables successfully, when I try to access them through the Jupyter Notebook I am not able to access them. However, I m able to access them through the shell inside the docker container.

Can anyone help me with this issue

**Comment by kevin-bates** [comment]
The set of environment variables available to a running notebook are captured at the time the notebook is opened (i.e., when the underlying kernel process is started).  These variables come directly from the Notebook Server process (as well as any configured in the `env:` stanza of the kernelspec json file) which are _static once the Notebook Server process starts_.

>Although I am able to set the variables successfully

It's not clear to me what this step consists of but I suspect the Notebook Server process has already started.

Unless there's something else going on, you'll need to start your docker container using `docker run ... -e ENV1=val1 -e ENV2=val2 ...` to convey user-specifc envs to the Notebook Server instance, which then flow to the kernel process.

**Comment by IamViditAgarwal** [comment]
> The set of environment variables available to a running notebook are captured at the time the notebook is opened (i.e., when the underlying kernel process is started). These variables come directly from the Notebook Server process (as well as any configured in the `env:` stanza of the kernelspec json file) which are _static once the Notebook Server process starts_.
> 
> > Although I am able to set the variables successfully
> 
> It's not clear to me what this step consists of but I suspect the Notebook Server process has already started.
> 
> Unless there's something else going on, you'll need to start your docker container using `docker run ... -e ENV1=val1 -e ENV2=val2 ...` to convey user-specifc envs to the Notebook Server instance, which then flow to the kernel process.

Isnt there a way where I can reflect the new env variables , without restarting the docker container. Bcz In my setting, I cant just restart the docker container with new values. I have to do in the running container only

**Comment by kevin-bates** [comment]
Not really.  I suppose you could mount the directory where the kernelspec (kernel.json) file is located and create/modify the file's `env:` stanza (or edit that file via `docker exec`).  Those variables would then apply to _next_ launch of that particular kernel.  But that's quite a hack.

Sounds like you could use a form of _paramaterized kernels_ where certain values are provided on a per-kernel basis.  These do not exist (to my knowledge), but there's been momentum building for this kind of feature. We should make sure environment variables are part of that effort.

FWIW, env variables are fairly static in nature since they are typically 'set' once the process starts.  As a result, they aren't a very good fit for dynamic behaviors within a single application.  Can you explain your use case a bit further?  Perhaps there are other alternatives.

**Comment by IamViditAgarwal** [comment]
> Not really. I suppose you could mount the directory where the kernelspec (kernel.json) file is located and create/modify the file's `env:` stanza (or edit that file via `docker exec`). Those variables would then apply to _next_ launch of that particular kernel. But that's quite a hack.
> 
> Sounds like you could use a form of _paramaterized kernels_ where certain values are provided on a per-kernel basis. These do not exist (to my knowledge), but there's been momentum building for this kind of feature. We should make sure environment variables are part of that effort.
> 
> FWIW, env variables are fairly static in nature since they are typically 'set' once the process starts. As a result, they aren't a very good fit for dynamic behaviors within a single application. Can you explain your use case a bit further? Perhaps there are other alternatives.


Let me explain you my case:
So , In my scenario: THere is a variable called TOKEN  whose value is set as the environment variable  when the docker starts. But the value of this TOKEN variable gets expired after a hour. So, I have made a service that runs in the background of the docker to update the value of TOKEN. Now, the issue is when I update the value of a TOKEN variable in background. Jupyter Notebook is not able to get the updated value. It is taking the older value.

This is my whole issue

**Comment by kevin-bates** [comment]
I see, thank you.  So this TOKEN value needs to be _current_ and reside in the environment of the kernel - which requires the current value to perform its operations?

Assuming that's correct.  One approach would be to have your background service instead write the updated TOKEN value to a file (local in the container) and modify the kernelspec to instead invoke a shell script that source's the file prior to launching the kernel.

For example, the service writes the following to `/path/to/token_file`: `TOKEN=my-token-for-use-in-the-kernel`.  Then, the kernel.json file for your kernel is modified to something like the following (I've added an `env:` stanza for example purposes and changed the display name that would appear in the Notebook UI):
```json
{
 "argv": [
  "/usr/local/share/jupyter/kernels/python3_token/run.sh",
  "/opt/anaconda3/envs/notebook-dev/bin/python",
  "-m",
  "ipykernel_launcher",
  "-f",
  "{connection_file}"
 ],
 "env": {
  "MY_PER_KERNEL_ENV": "MY_PER_KERNEL_VALUE"
 },
 "display_name": "Python 3 with token",
 "language": "python"
}
```
I also assume your `kernel.json` is located in `/usr/local/share/jupyter/kernels/python3_token`.  Then `run.sh` would resemble the following...
```bash
#!/bin/bash
source /path/to/token_file
$* 
```
I hope that helps.

**Comment by kevin-bates** [comment]
Just noticed that if you're going to use `source` you'll probably want the file entries to include `export` prior to the assignment, else it doesn't look like its available to the kernel.  Note also that this could be a general set of envs if there are others you need (and you're using the `source` approach).

**Comment by IamViditAgarwal** [comment]
> I see, thank you. So this TOKEN value needs to be _current_ and reside in the environment of the kernel - which requires the current value to perform its operations?
> 
> Assuming that's correct. One approach would be to have your background service instead write the updated TOKEN value to a file (local in the container) and modify the kernelspec to instead invoke a shell script that source's the file prior to launching the kernel.
> 
> For example, the service writes the following to `/path/to/token_file`: `TOKEN=my-token-for-use-in-the-kernel`. Then, the kernel.json file for your kernel is modified to something like the following (I've added an `env:` stanza for example purposes and changed the display name that would appear in the Notebook UI):
> 
> ```json
> {
>  "argv": [
>   "/usr/local/share/jupyter/kernels/python3_token/run.sh",
>   "/opt/anaconda3/envs/notebook-dev/bin/python",
>   "-m",
>   "ipykernel_launcher",
>   "-f",
>   "{connection_file}"
>  ],
>  "env": {
>   "MY_PER_KERNEL_ENV": "MY_PER_KERNEL_VALUE"
>  },
>  "display_name": "Python 3 with token",
>  "language": "python"
> }
> ```
> 
> I also assume your `kernel.json` is located in `/usr/local/share/jupyter/kernels/python3_token`. Then `run.sh` would resemble the following...
> 
> ```shell
> #!/bin/bash
> source /path/to/token_file
> $* 
> ```
> 
> I hope that helps.

Ohk , Thanks .. I will try your solution once and will let u know
